pkg/kubelet/metrics/collectors/cri_metrics.go

```diff
@@ -0,0 +1,123 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/component-base/metrics"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/klog/v2"
+)
+
+type criMetricsCollector struct {
+	metrics.BaseStableCollector
+	descriptors           map[string]*metrics.Desc
+	listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error)
+}
+
+// Check if criMetricsCollector implements necessary interface
+var _ metrics.StableCollector = &criMetricsCollector{}
+
+// NewCRIMetricsCollector implements the metrics.Collector interface
+func NewCRIMetricsCollector(ctx context.Context, listPodSandboxMetrics func(context.Context) ([]*runtimeapi.PodSandboxMetrics, error), listMetricDescriptors func(context.Context) ([]*runtimeapi.MetricDescriptor, error)) metrics.StableCollector {
+	descs, err := listMetricDescriptors(ctx)
+	if err != nil {
+		klog.ErrorS(err, "Error reading MetricDescriptors")
+		return &criMetricsCollector{
+			listPodSandboxMetrics: listPodSandboxMetrics,
+		}
+	}
+	c := &criMetricsCollector{
+		listPodSandboxMetrics: listPodSandboxMetrics,
+		descriptors:           make(map[string]*metrics.Desc, len(descs)),
+	}
+
+	for _, desc := range descs {
+		c.descriptors[desc.Name] = criDescToProm(desc)
+	}
+
+	return c
+}
+
+// Describe implements the metrics.DescribeWithStability interface.
+func (c *criMetricsCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	for _, desc := range c.descriptors {
+		ch <- desc
+	}
+}
+
+// Collect implements the metrics.CollectWithStability interface.
+// TODO(haircommander): would it be better if these were processed async?
+func (c *criMetricsCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	podMetrics, err := c.listPodSandboxMetrics(context.Background())
+	if err != nil {
+		klog.ErrorS(err, "Failed to get pod metrics")
+		return
+	}
+
+	for _, podMetric := range podMetrics {
+		for _, metric := range podMetric.GetMetrics() {
+			promMetric, err := c.criMetricToProm(metric)
+			if err == nil {
+				ch <- promMetric
+			}
+		}
+		for _, ctrMetric := range podMetric.GetContainerMetrics() {
+			for _, metric := range ctrMetric.GetMetrics() {
+				promMetric, err := c.criMetricToProm(metric)
+				if err == nil {
+					ch <- promMetric
+				}
+			}
+		}
+	}
+}
+
+func criDescToProm(m *runtimeapi.MetricDescriptor) *metrics.Desc {
+	return metrics.NewDesc(m.Name, m.Help, m.LabelKeys, nil, metrics.INTERNAL, "")
```

Comment:
nit: maybe worth to note here, the label keys are "variableLabels" as compared to "constant labels" (which is passed as nil), because the values of the labels will be different for each container (xref: https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/prometheus/client_golang/prometheus/desc.go#L77-L81)